### PR TITLE
Uqbar Rollup Testnet: %sequencer rewrite, remove %ziggurat, create %uqbar: read/write "zane"

### PR DESCRIPTION
**Pull request by dr-frmr**
This PR removes the consensus engine that was `%ziggurat`. It replaces consensus with a rollup contract simulator, which is the `%rollup` agent. 

We've also added `%uqbar`, an agent that acts as a vane for Uqbar interactions. 

Finally, there are a number of changes to the execution engine that have made their way into this PR. `mill.hoon` now emits a proper state transition, including hints for Zink. Town IDs now match the rest of the chain in being `@ux`. Contracts have a changed context, where `now` has replaced `blocknum`.

Mill still needs to be parallelized, which will entail restructuring the `mill-all` arm to be a big `turn` over the list of txns, and at that point we can get the `diff` granary we need to provide to the rollup.

##  Uqbar.hoon

```
      (a)       (b)            (c)
Users <--> Zane --> Sequencers --> Rollup
             ^      |  (e)        |
         (d) |      V             | (f)
             Indexers   <----------
```
Basic APIs / interfaces

(a) Users <--> Zane
- Gall dApps use zane to query their contract address, perform reads, and submit transactions (signed via wallet)
- Zane allows apps to track selected grains and subscribe to get those changes

(b) Zane --> Sequencers
- Zane submits transactions directly to sequencer (it gets the information to do so from Indexer)

(c) Sequencers --> Rollup
- Sequencers post batches to rollup
- Batches contain everything needed for ZK-verifiable state transitions

(d) Indexers --> Zane
- Zane chooses a ship hosting an indexer to serve as a data source
- Indexer provides requested information along with mapping of town to sequencer (so zane knows who to submit to)

(e) Sequencers --> Indexers
- Indexers subscribe to one or more sequencers to get latest state (note that batches have replaced blocks, but same concept)
- Indexers verify received state against state roots posted to rollup

(f) Rollup --> Indexers
- Indexers subscribe to the rollup contract (currently simulated by a gall agent) to get latest state roots for desired towns, so they can verify state received by sequencers

---

##  Uqbar Rollup Testnet Spec

Goal: take out all consensus machinery from ziggurat and sequencer apps, and replace them with a unified rollup sequencer designed off this spec: web+urbitgraph://group/~datwet/uqteam/graph/~datwet/dev-posts-685/170141184505683749127804303000964431872
(spec pasted here for posterity, can edit as implemented):

Because Uqbar is currently designed as a rollup, in order to run a testnet for it, we need a system that it can "roll up" to. While we could write a Solidity contract and deploy on an Eth testnet, that would require substantial time, and also not generalize if we want to deploy our rollup contract on an L2.

Instead, we will construct an Urbit program that simulates a rollup contract. This will
- allow users to run transactions on our testnet
- let us get practical experience coordinating multiple towns
- provide a framework to experiment quickly with settlement and protocol features
- be a reference implementation off which we base our rollup smart contract(s)

###  high-level spec
- multiple towns
- each town has one sequencer, defined by a pubkey
- set of ships (multisig) controls sequencer assigned to each town These ships can be called "protocol operators"
- each town maintains a data structure containing:
- data availability method (DAC or full publish)
- append-only list of state diffs (if full publish). Simulates passing diffs as call data
- hash of state diff regardless of data avail method used
- append-only list of state roots
- current allowed sequencer

### CODE REVIEW

**Comment by dr-frmr** [comment]
Basic machinery is finished, now need to adjust mill to produce the diffs we want. After that, can fully remove ziggurat and use this!

**Comment by dr-frmr** [comment]
This will require a bunch of little changes to both wallet and indexer. Nothing huge though. I will make wallet changes.

**Comment by dr-frmr** [comment]
SO we need `mill-all` to produce, instead of our old `chunk` def, 4 things:

- the new state (pair of granary (map of grains) and populace (nonces))
- the JSON hint for ZK proving of the transition
- a "state diff", which is a granary that contains all the changed and issued grains from the txn batch
- the list of events from the txn batch (all generated `crow`s)

**Comment by tacryt-socryp** [comment]
Reasonable. So does any individual sequencer only get to know about its particular shard? How would cross-shard transfers work?

**Comment by dr-frmr** [comment]
> Reasonable. So does any individual sequencer only get to know about its particular shard? How would cross-shard transfers work?

The transfers will have to be initiated by a sequencer on the starting town post a 'burn' with the grains involved, proving they're removed from state in that batch and indicating a desired destination town. I guess we need another field in the state transition type to hold these 'burns'. Then the receiving town can read this (either directly from the sequencer or from the rollup once the batch is posted?) and instantiate those grains in its own state as part of its next batch.

**Comment by dr-frmr** [comment]
That’s part of what the peer-root system is for — you make every new batch on a town commit to recent-enough roots to every other active town. (How recent is recent enough, I’m not yet sure..)

**Comment by hosted-fornet** [comment]
In 5d5d839 the `%indexer-update` has a `town` which has `grain`s, but no `egg`s. In the current system with `ziggurat` we can also display `egg`s, which I think is valuable. Can the `%indexer-update` also include the `basket` for a given batch? Regarding pending transactions, you could also imagine adding a scry that returns the current `basket`, though I'm not certain this is the right way to make this info available to the frontend.

**Comment by dr-frmr** [comment]
Agreed that I shouldn’t have taken the transactions out of the update type — I’ll put those back in @hosted-fornet

**Comment by dr-frmr** [comment]
For pending transactions it will ultimately be a choice made by the sequencer to permit visibility to their private mempool

**Comment by dr-frmr** [comment]
Brought in latest from master by way of Logan's mill hint issuance, compiling. @hosted-fornet I changed a minimal amount of types in `indexer-bowl.hoon` just to make everything compile

**Comment by dr-frmr** [comment]
@tacryt-socryp I'm getting an empty result when using `zebra` to run contracts, when `mule` is working. Do you think I need to recompute the hash-cache, since I recompiled smart-lib?

**Comment by tacryt-socryp** [comment]
Definitely running out of gas. I'll put a new hash cache in there.

**Comment by dr-frmr** [comment]
> Definitely running out of gas. I'll put a new hash cache in there.

I got it! Added a new hash cache and ran some stuff through zebra, which was awesome. Just takes more gas and more time.

**Comment by hosted-fornet** [comment]
Having trouble running this with the jetted Pederson hash. Here's the process I am following:

1. Compile binary from the following branch: https://github.com/martyr-binbex/urbit/tree/pedersen-hash-jet
2. Run my fakezod using that binary.
3. Bring the `zuse.hoon` from that branch into my fakezod's `base` desk and commit it.
4. Bring the code here into `zig` desk and commit, install, and run via:
```
|install our %zig
:rollup|activate
:sequencer|init our 0x0 0xf3b2.f5ab.92df.a29d.5e8e.fe70.98cb.ed99.3856.949b.fb4f.5cfb.7bca.4a45.0ae7.0e50
:indexer &set-sequencer [our %sequencer]
:indexer &set-rollup [our %rollup]
:uqbar|set-sources our 0x0 our
:wallet &zig-wallet-poke [%populate 0xbeef]
:wallet &zig-wallet-poke [%submit-custom from=0x7772.b8a7.6840.8922.2903.5b28.7494.436f.8850.713c to=0x74.6361.7274.6e6f.632d.7367.697a town=0x0 gas=[1 1.000.000] args='[%give 0xc7ec.a38c.5c74.d58d.04b0.6650.4772.f3a6.d02e.92f5 `0x532c.d5cf.befc.5c0f.e5dd.3910.8f23.79fd 777 1.000]' my-grains=(silt ~[0x532c.d5cf.befc.5c0f.975c.8e60.c3f6.fcac]) cont-grains=(silt ~[0x532c.d5cf.befc.5c0f.e5dd.3910.8f23.79fd])]
:sequencer &sequencer-town-action [%trigger-batch ~]
```

The `%trigger-batch` spins for a long time before eventually outputting a jet mismatch-type error message and then `%dec` for a long time, like:
```
fund: in by, parent 18226962 not found at 7
%dec
%dec                
...                                               
```

Recompiling smart-lib and contracts did not lead to any diffs in those binaries.

What am I doing wrong here?

**Comment by martyr-binbex** [comment]
> Having trouble running this with the jetted Pederson hash. Here's the process I am following:
> 
> 1. Compile binary from the following branch: https://github.com/martyr-binbex/urbit/tree/pedersen-hash-jet
> 2. Run my fakezod using that binary.
> 3. Bring the `zuse.hoon` from that branch into my fakezod's `base` desk and commit it.
> 4. Bring the code here into `zig` desk and commit, install, and run via:
> 
> ```
> |install our %zig
> :rollup|activate
> :sequencer|init our 0x0 0xf3b2.f5ab.92df.a29d.5e8e.fe70.98cb.ed99.3856.949b.fb4f.5cfb.7bca.4a45.0ae7.0e50
> :indexer &set-sequencer [our %sequencer]
> :indexer &set-rollup [our %rollup]
> :uqbar|set-sources our 0x0 our
> :wallet &zig-wallet-poke [%populate 0xbeef]
> :wallet &zig-wallet-poke [%submit-custom from=0x7772.b8a7.6840.8922.2903.5b28.7494.436f.8850.713c to=0x74.6361.7274.6e6f.632d.7367.697a town=0x0 gas=[1 1.000.000] args='[%give 0xc7ec.a38c.5c74.d58d.04b0.6650.4772.f3a6.d02e.92f5 `0x532c.d5cf.befc.5c0f.e5dd.3910.8f23.79fd 777 1.000]' my-grains=(silt ~[0x532c.d5cf.befc.5c0f.975c.8e60.c3f6.fcac]) cont-grains=(silt ~[0x532c.d5cf.befc.5c0f.e5dd.3910.8f23.79fd])]
> :sequencer &sequencer-town-action [%trigger-batch ~]
> ```
> 
> The `%trigger-batch` spins for a long time before eventually outputting a jet mismatch-type error message and then `%dec` for a long time, like:
> 
> ```
> fund: in by, parent 18226962 not found at 7
> %dec
> %dec                
> ...                                               
> ```
> 
> Recompiling smart-lib and contracts did not lead to any diffs in those binaries.
> 
> What am I doing wrong here?

**Comment by martyr-binbex** [comment]
One thing is you need to update all the references to hash.pedersen in lib/zink/zink.hoon to hash.pedersen.secp.crypto

**Comment by hosted-fornet** [comment]
Okay, doing so I no longer get the initial `fund` output, but I still get `%dec`s for multiple minutes before I Ctrl-C.

**Comment by tacryt-socryp** [comment]
Axis matching for jets with the new libraries still needs to be tested and worked out.

**Comment by hosted-fornet** [comment]
Cool, I will just go back to mule for testing for now, then.

**Comment by hosted-fornet** [comment]
I'll be sending a few bugfix commits here momentarily.

**Comment by hosted-fornet** [comment]
Okay, brought in the lib-i-fied contracts from master and fixed an indexer bug where historical grains returned by the `/json/grain` scry would give the improper `data` (they would all give the most recent state for their data, rather than historical data, because they weren't passing in that historical data).

Bringing in the lib-i-fied contracts will likely cause merge conflicts on branches pointing to this one, so sorry about that.

**Comment by hosted-fornet** [comment]
The rollup now sends a timestamp on `%new-peer-root` to establish a canonical time for a given batch. Previously, indexers on different ships would hackily timestamp a batch upon receipt, leading to different timestamps for the same batch depending on which indexer you query.

**Review by hosted-fornet — APPROVED** [review]

**Comment by hosted-fornet** [comment]
To test on multiple ships I did:

```
::  on `~zod`
.smart-lib/noun +zig!mk-smart
|install our %zig
:rollup|activate
:sequencer|init our 0x0 0xf3b2.f5ab.92df.a29d.5e8e.fe70.98cb.ed99.3856.949b.fb4f.5cfb.7bca.4a45.0ae7.0e50
:indexer &set-sequencer [our %sequencer]
:indexer &set-rollup [our %rollup]
:uqbar|set-sources our 0x0 our
:wallet &zig-wallet-poke [%populate 0xbeef]

::  AFTER `~zod` IS DONE, on other ship (say, `~nec`)
.smart-lib/noun +zig!mk-smart
|install our %zig
:indexer &set-sequencer [~zod %sequencer]
:indexer &set-rollup [~zod %rollup]
:uqbar|set-sources our 0x0 our

::  AFTER `~nec` is done, on `~zod`
:wallet &zig-wallet-poke [%submit-custom from=0x7772.b8a7.6840.8922.2903.5b28.7494.436f.8850.713c to=0x74.6361.7274.6e6f.632d.7367.697a town=0x0 gas=[1 1.000.000] args='[%give 0xc7ec.a38c.5c74.d58d.04b0.6650.4772.f3a6.d02e.92f5 `0x532c.d5cf.befc.5c0f.e5dd.3910.8f23.79fd 10.000.000.000.000.000.000 1.000]' my-grains=(silt ~[0x532c.d5cf.befc.5c0f.975c.8e60.c3f6.fcac]) cont-grains=(silt ~[0x532c.d5cf.befc.5c0f.e5dd.3910.8f23.79fd])]
:sequencer &sequencer-town-action [%trigger-batch ~]
```

You can then query state using the indexer as follows (assuming `~zod` is at `localhost:8080` and `~nec` is at `localhost:8081`):

```bash
# in bash, set cookies
export ZOD_COOKIE=$(curl -i -X POST localhost:8080/~/login -d 'password=lidlut-tabwed-pillex-ridrup' | grep set-cookie | awk '{print $2}' | awk -F ';' '{print $1}')
export NEC_COOKIE=$(curl -i -X POST localhost:8081/~/login -d 'password=ropnys-batwyd-nossyt-mapwet' | grep set-cookie | awk '{print $2}' | awk -F ';' '{print $1}')

# query `~zod`, including `grain` `data`
curl --cookie "$ZOD_COOKIE" localhost:8080/~/scry/indexer/json/grain/0x532c.d5cf.befc.5c0f.975c.8e60.c3f6.fcac.json | jq

# query `~nec`, NOT including `grain` `data` (which currently requires a locally-run sequencer)
curl --cookie "$NEC_COOKIE" localhost:8081/~/scry/indexer/grain/0x532c.d5cf.befc.5c0f.975c.8e60.c3f6.fcac.json | jq
```

Timestamps should be the same, as should the number of historical instances of the rice.

**Comment by hosted-fornet** [comment]
Ah, and testing was using mule, not zebra.

File: app/wallet.hoon [hunk 1]
```diff
@@ -79,53 +77,57 @@
...
     ::
         %derive-new-address
       ::  if hdpath input is empty, use address-index+1 to get next
       =+  (from-seed:bip32 [64 (to-seed:bip39 (trip mnem.seed.state) (trip pass.seed.state))])
-      =+  %-  derive-path:-
-          ?:  !=("" hdpath.act)  hdpath.act
-          (weld "m/44'/60'/0'/0/" (scow %ud address-index.seed.state))
-      :-  (create-holder-and-id-subs (silt ~[public-key:-]) (need indexer.state))
+      =/  core
+        %-  derive-path:-
```
Comment (tacryt-socryp) [inline app/wallet.hoon:119]: Using `-` here is a little nasty, give cores names.

File: app/wallet.hoon [hunk 2]
```diff
@@ -198,17 +189,16 @@
           ==
       :~  (tx-update-card egg.p `args.p)
           :*  %pass  /submit-tx/(scot %ux egg-hash)
-              %agent  [node ?:(=(0 town-id.p.egg.p) %ziggurat %sequencer)]
-              %poke  %zig-weave-poke
-              !>([%forward (silt ~[egg.p])])
+              %agent  [our.bowl %uqbar]
+              %poke  %uqbar-write
+              !>([%submit egg.p])
```
Comment (tacryt-socryp) [inline app/wallet.hoon:194]: Cast all inputs to vases to prevent type mismatches in the future.

File: app/wallet.hoon [hunk 3]
```diff
@@ -246,9 +236,9 @@
           ==
       :~  (tx-update-card egg `[%custom args.act])
           :*  %pass  /submit-tx/(scot %ux from.act)/(scot %ux egg-hash)
-              %agent  [node ?:(=(0 town.act) %ziggurat %sequencer)]
-              %poke  %zig-weave-poke
-              !>([%forward (silt ~[egg])])
+              %agent  [our.bowl %uqbar]
+              %poke  %uqbar-write
+              !>([%submit egg])
```
Comment (tacryt-socryp) [inline app/wallet.hoon:241]: Same

File: app/wallet.hoon [hunk 4]
```diff
@@ -335,9 +317,9 @@
           ==
       :~  (tx-update-card egg `args.act)
           :*  %pass  /submit-tx/(scot %ux from.act)/(scot %ux egg-hash)
-              %agent  [node ?:(=(0 town.act) %ziggurat %sequencer)]
-              %poke  %zig-weave-poke
-              !>([%forward (silt ~[egg])])
+              %agent  [our.bowl %uqbar]
+              %poke  %uqbar-write
+              !>([%submit egg])
```
Comment (tacryt-socryp) [inline app/wallet.hoon:322]: Same

File: app/indexer.hoon [hunk 5]
```diff
@@ -133,647 +154,722 @@
...
-        %^  consume-slot  epoch-num.update
-        epoch-start-time.update  [header.update blk.update]
+          %new-peer-root
+        =*  town-id  town-id.update
+        ?~  town-q=(~(get by sequencer-update-queue) town-id)
+          :-  ~
+          %=  state
+              town-update-queue
+            %+  ~(put by town-update-queue)  town-id
+            %+  %~  put  by
```
Comment (tacryt-socryp) [inline app/indexer.hoon:463]: neater to use `+fall` than handroll this

File: app/rollup.hoon [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+        !!
+      ::  handle full-publish mode
+      ::
+      =+  %=  u.hall
+            latest-diff-hash  diff-hash.act
+            roots  (snoc roots.u.hall new-root.act)
+          ==
+      =+  (~(put by capitol) town-id.act -)
+      :_  state(capitol -)
+      :~  [%give %fact ~[/peer-root-updates] %sequencer-rollup-update !>(`town-update`[%new-peer-root town-id.act new-root.act now.bowl])]
```
Comment (tacryt-socryp) [inline app/rollup.hoon:129]: line crazy long

File: app/rollup.hoon [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+      `state(status %available)
+    ::
+        %launch-town
+      ::  create new hall
+      ?<  (~(has by capitol) town-id.hall.act)
+      ::  TODO remove starting-state from init and populate new towns via
+      ::  assets from other towns
+      =+  (~(put by capitol) town-id.hall.act hall.act)
+      :_  state(capitol -)
+      :~  [%give %fact ~[/peer-root-updates] %sequencer-rollup-update !>(`town-update`[%new-peer-root town-id.hall.act (rear roots.hall.act) now.bowl])]
```
Comment (tacryt-socryp) [inline app/rollup.hoon:82]: long

File: app/sequencer.hoon [hunk 8]
```diff
@@ -31,215 +34,230 @@
...
-      ?.  (~(has by wex.bowl) [/sequencer/updates our.bowl %ziggurat])
-        ~[(sequencer-sub-card our.bowl)]
-      ~
+      ?>  =(src.bowl our.bowl)
+      ?.  =(%off status.state)
+        ~|("%sequencer: already active" !!)
+      ::  poke rollup ship with params of new town
+      ::  (will be rejected if id is taken)
+      =/  =land  ?~(starting-state.act [~ ~] u.starting-state.act)
+      =/  new-root  (shax (jam land))
```
Comment (tacryt-socryp) [inline app/sequencer.hoon:98]: No reason to make a hash of a jam. Just make a hash of the noun directly.

File: app/rollup.hoon (hunk 6 above)
Comment (dr-frmr) [reply app/rollup.hoon:129]:
```
:~  =-  [%give %fact ~[/peer-root-updates] %sequencer-rollup-update -]
          !>(`town-update`[%new-peer-root town-id.act new-root.act now.bowl])
      ::
          =-  [%give %fact ~[/capitol-updates] %sequencer-rollup-update -]
          !>(`capitol-update`[%new-capitol -])
      ==
```
How do you feel about something like this?

Comment (dr-frmr) [reply app/rollup.hoon:129]: (ignore spacing, github codeblock being bad)

File: app/sequencer.hoon (hunk 8 above)
Comment (dr-frmr) [reply app/sequencer.hoon:98]: How do I hash a noun directly? I know `sham` allows for noun inputs but doesn't it just use jam internally? Is that jetted more helpfully?

File: app/rollup.hoon (hunk 6 above)
Comment (tacryt-socryp) [reply app/rollup.hoon]: Much better

File: app/sequencer.hoon (hunk 8 above)
Comment (tacryt-socryp) [reply app/sequencer.hoon:98]: Use `+sham` so as to include the logic for salting differently to differentiate atom case and cell case. It also won't unnecessarily `+jam` if input is an atom. In any case, no need to handroll the functionality yourself. Probably not a performance difference though now that I look closer

**Comment by tacryt-socryp** [comment]
LGTM

**Pull request merged**